scan-task.ts

```diff
@@ -1,92 +1,99 @@
 }
-export const scan = (siteUrl: string, opts: ScanOptions) => {
-  const dir = opts.dataDirectory;
+
+type ScanEvents = {
+  warning: (message: string | Error) => void;
+  info: (message: string) => void;
+  reportComplete: (url: string, reportData: string) => void;
+  urlFound: (
+    url: string,
+    contentType: string,
+    bytes: number,
+    statusCode: number
+  ) => void;
+};
+
+export const scan = (
+  siteUrl: string,
+  {
+    crawler = defaultCrawler,
+    lighthouse = runLighthouseReport,
+    dataDirectory,
+    shouldRunLighthouseOnURL,
+    ...opts
+  }: ScanOptions
+) => {
+  const { promise, on, emit } = createEmitter<ScanEvents>();
   // Set up for lighthouse reports
-  const reportFormat = 'csv'; // Html works too
   const reportDirName = 'reports';
-  const reportsDirPath = `${dir}/${reportDirName}`;
-  fs.mkdirSync(reportsDirPath, { recursive: true });
+  const reportsDirPath = `${dataDirectory}/${reportDirName}`;
 
-  // Set up for crawler
-  const crawler = new Crawler(siteUrl);
-  if (opts.userAgent) crawler.userAgent = opts.userAgent;
-  crawler.respectRobotsTxt = !opts.ignoreRobotsTxt;
-  fs.mkdirSync(dir, { recursive: true });
-  const file = `${dir}/urls.csv`;
-  fs.writeFileSync(file, 'URL,content_type,bytes,response\n', {
-    encoding: 'utf-8',
-  });
+  fs.mkdirSync(dataDirectory, { recursive: true });
   /** Used so we can display an error if no pages are found while crawling */
   let hasFoundAnyPages = false;
-  console.log('Created CSV file');
-  const stream = fs.createWriteStream(file, { flags: 'a' });
-  crawler.on('fetchcomplete', (queueItem, responseBuffer, response) => {
-    hasFoundAnyPages = true;
-    console.log(
-      'Crawled %s [%s] (%d bytes)',
-      queueItem.url,
-      response.headers['content-type'],
-      responseBuffer.length
-    );
-    stream.write(makeUrlRow(queueItem, responseBuffer, response));
-    const reportFileName = makeFileNameFromUrl(queueItem.url, reportFormat);
-    if (!fileDoesntExist(reportFileName, reportsDirPath)) {
-      console.log('Skipping report because file already exists');
-      return;
-    }
 
-    if (!isContentTypeHtml(response.headers['content-type'])) {
-      return;
-    }
+  emit('info', 'Starting the crawl...');
+
+  const crawlerEmitter = crawler(siteUrl, opts);
 
-    const reportData = runReport(queueItem.url, reportFormat);
-    if (!reportData) return;
-    fs.writeFileSync(`${reportsDirPath}/${reportFileName}`, reportData);
-    console.log(`Wrote report for ${queueItem.url}`);
+  crawlerEmitter.on('urlFound', (url, contentType, bytes, statusCode) => {
+    hasFoundAnyPages = true;
+    // TODO in code review: move this console.log up to cli
+    console.log('Crawled %s [%s] (%d bytes)', url, contentType, bytes);
+    if (!isContentTypeHtml(contentType)) return;
+    emit('urlFound', url, contentType, bytes, statusCode);
+    // TODO in code review: should non-HTML URL's be written to the urls.csv and logged or not?
```

Comment:
@emersonthis what are your thoughts on skipping logging and writing to `urls.csv` for URLs that aren't `text/html`?

Reply:
I'm not sure I completely understand the question, but I was thinking it would eventually be optimal if we could configure the crawler to only discover HTML pages. Just skip the assets entirely.

Reply:
Right now non-HTML files still get added to `urls.csv` and they still get logged. My question was if I can remove those two features in this PR.

I agree that eventually we may want to configure the crawler to skip them, but for now is it fine to just suppress them from the output?